# Application of the Markov Chain to calculate the probability of a state 
# transition from Bear Market to Bull Market in Vietnam stock market

The Theory of the Markov Chain, readers can read it here "https://en.wikipedia.org/wiki/Markov_chain"

I assume that state 1 of the market is a bull market for a stock index that is more than 200 session average, and state 2 of the market is a bear market for a stock index that is 200 sessions smaller than the average.

The data used is the Vietnam stock market index from 2013 to 2023. Readers can view it here
https://drive.google.com/file/d/1haF-S_W5codfcILa9AtiXhOr8L-fn4HI/view

In [1]:
import pandas as pd
import numpy as np

In [5]:
# load data
data = pd.read_csv(r"C:\Users\DPC\Desktop\data\markov.csv")
print(len(data))
data.head()

2297


date   price    open    high     low      vol  change      ma200
0   12/9/2013  511.23  514.13  514.13  510.94  102.31K  0.0022  492.89210
1  12/10/2013  511.60  512.10  513.83  510.24   80.98K  0.0007  493.02980
2  12/11/2013  505.31  511.74  511.81   504.9  103.54K -0.0123  493.26255
3  12/12/2013  506.92  503.77  507.68  503.76   80.16K  0.0032  493.46050
4  12/13/2013  506.06  507.07  509.35  505.49   67.67K -0.0017  493.62230

In [12]:
# Set condition: if price > moving average 200 then 1 else 0
condition = []
for i in range(len(data)):
    if data.iloc[i,1] > data.iloc[i,7]:
        condition.append(1)
    else:
        condition.append(0)
data["condition"] = condition
data.head()

date   price    open    high     low      vol  change      ma200  \
0   12/9/2013  511.23  514.13  514.13  510.94  102.31K  0.0022  492.89210   
1  12/10/2013  511.60  512.10  513.83  510.24   80.98K  0.0007  493.02980   
2  12/11/2013  505.31  511.74  511.81   504.9  103.54K -0.0123  493.26255   
3  12/12/2013  506.92  503.77  507.68  503.76   80.16K  0.0032  493.46050   
4  12/13/2013  506.06  507.07  509.35  505.49   67.67K -0.0017  493.62230   

   condition  
0          1  
1          1  
2          1  
3          1  
4          1

In [42]:
# Transition from state 1 to state 1
t11 = []
for i in range(len(data)-1):
    if data.iloc[i,8] == 1:
        if data.iloc[i+1,8] == 1:
            t11.append(1)
        else: 
            t11.append(0)
    else: 
        t11.append(0)
sumt11 = sum(t11)
print(sumt11)

1446


In [43]:
# Transition from state 1 to state 0
t10 = []
for i in range(len(data)-1):
    if data.iloc[i,8] == 1:
        if data.iloc[i+1,8] == 0:
            t10.append(1)
        else: 
            t10.append(0)
    else: 
        t10.append(0)
sumt10 = sum(t10)
print(sumt10)

22


In [44]:
# Probability of state transition 1 to 1
p11 = sumt11 / (sumt11+sumt10)
# Probability of state transition 1 to 0
p10 = 1 - p11
print("p11 =",p11)
print("p10 =",p10)

p11 = 0.9850136239782016
p10 = 0.01498637602179842


In [41]:
# Transition from state 0 to state 0
t00 = []
for i in range(len(data)-1):
    if data.iloc[i,8] == 0:
        if data.iloc[i+1,8] == 0:
            t00.append(1)
        else: 
            t00.append(0)
    else: 
        t00.append(0)
sumt00 = sum(t00)
print(sumt00)

807


In [45]:
# Transition from state 0 to state 1
t01 = []
for i in range(len(data)-1):
    if data.iloc[i,8] == 0:
        if data.iloc[i+1,8] == 1:
            t01.append(1)
        else: 
            t01.append(0)
    else: 
        t01.append(0)
sumt01 = sum(t01)
print(sumt01)

21


In [47]:
# Probability of state transition 0 to 0
p00 = sumt00 / (sumt00+sumt01)
# Probability of state transition 0 to 1
p01 = 1 - p00
print("p11 =",p00)
print("p10 =",p01)

p11 = 0.9746376811594203
p10 = 0.025362318840579712


In [48]:
# Transition Probability Matrix
markov_matrix = np.array([[p11,p01],[p10,p00]])
print(markov_matrix)

[[0.98501362 0.02536232]
 [0.01498638 0.97463768]]


In [49]:
#The long run state of the Markov chain
from numpy.linalg import eig
w,v=eig(markov_matrix)
print('E-value:', w)
print('E-vector', v)

E-value: [1.         0.95965131]
E-vector [[ 0.8609332  -0.70710678]
 [ 0.50871802  0.70710678]]


#### In the long term, the probability vector [ p(state1/state1), p(state1/state0)] will be the eigenvector of the Transition Probability
#### Matrix. In this case, I would choose p(state1/state1) = 0.50871802 and p(state0/state1) = 0.70710678

# About the Authors:

My name is Nguyen Hung Truong, I am currently a research consultant at Worldquant investment LLC. My main job is to focus on studying patterns to come up with optimal strategies in stock trading.